### DISCUSSION TITLE
[Feature Request] Table Panel Vertical Headers

### DISCUSSION THREAD

**Discussion by merwy** [grafana/grafana#36973] [Dashboards, Panels & Visualizations]
At first I want to say that we love Grafana with all of it’s fancy features.

I’m working in big organization which is going to utilize InfluxDB + Grafana on a large scale for monitoring thousands of servers.

We have an important use-case for using table panels in configuration as shown in below screenshot:
![table_issues](https://cloud.githubusercontent.com/assets/19647399/15887070/774f2152-2d60-11e6-9226-344ff76b0d4c.png)

The table is using **_Time series to columns_** option.

We are facing the following issues/inconveniences with current implementation:
1. Column headers/names are not wrapped which makes it impossible to see the column name text.
   ![table_bug1](https://cloud.githubusercontent.com/assets/19647399/15887200/36cd70ce-2d61-11e6-9801-2ada4421673e.png)
   **SOLUTION**: Provide column headers wrapping or option to display column headers vertically
2. Column headers are displayed even outside panel space. We have really a lot of columns and it seems Grafana is trying to display all of them despite of panel width. It seems to be a bug.
   ![table_bug2](https://cloud.githubusercontent.com/assets/19647399/15887250/8a9efe02-2d61-11e6-9028-7b8e69021be0.png)

Other improvement ideas:
1. Option do flip table by 90⁰ to have time o X-axis and columns (in our case rack names) on Y-axis.
   **MOTIVATION**: It is much more natural and intuitive having time on X-axis.
2. Possibility to define more than 2 thresholds, so that cell coloring gradation is more granular. 
   This would work then as kind of heatmap.

Currently we are using Influxdb 0.12.2 + Grafana 3.0.4.

We would really appreciate it if above improvements are implemented.
Thanks a lot!

**Comment by Pawcio4** [comment]
I agree. It would be really nice to have such improvements!

**Comment by torkelo** [comment]
none of these are bugs, you table is simply to big :) 

The column header layout issues somes from the fact hat the table supports fixed header (that stays in place when you scroll table). So solving the layout issues for the column headers could be impossible or tricky if the fixed header behavior should be kept. 

Try to split it into two table so that it fits horizontally.

**Comment by merwy** [comment]
Thanks for your comments. 
How about just having an option to wrap or put headers vertically? This would definatelly solve our issues. Please check mockup below to see how much more concent we would be able to fit into table if column width is smaller with vertical headers:
![proposal](https://cloud.githubusercontent.com/assets/19647399/15924740/f77738b4-2e33-11e6-94b9-23ea60757b54.png)

But actually, the possibility to rotate table 90 degrees as described in improvement 1 above in my initial post would be the best solution for us!

Please consider and comment also my 2 improvement ideas from my initial post. 

Thanks a lot!!!

**Comment by torkelo** [comment]
vertical headers sound like a very good idea!

**Comment by Pawcio4** [comment]
Vertical headers are good to have, that's true.
But looking at this sample formating:
![image](https://cloud.githubusercontent.com/assets/7103563/15929176/b9b5aa2a-2e4b-11e6-8f11-f0eeb44b8417.png)

Names of the servers that are outside of the table are displayed in 2 row format. Is there any way to do the same for headers inside table?

**Comment by merwy** [comment]
Thanks torkelo. In which release we can expect those vertical headers (or text wrapping) to be implemented in Grafana?

**Comment by torkelo** [comment]
@merwy it depends on when some submits a PR for the feature :)

**Comment by jdbranham** [comment]
@torkelo 
I was thinking about creating a heatmap by copying the table plugin, but should I consider modifying the existing table plugin with the vertical label options + gradient color mapping?
I've done something similar with the diagram plugin; using a canvas gradient with multiple color stops that correspond to an unbound color list. 

https://raw.githubusercontent.com/jdbranham/grafana-diagram/master/dist/img/diagram_gradient.PNG

**Comment by torkelo** [comment]
table panel? is not the graph panel more appropriate ?

**Comment by jdbranham** [comment]
I was thinking a grid based heatmap, but this also looks interesting - 
https://plot.ly/javascript/heatmaps/

**Comment by SeriousM** [comment]
This is an old issue, but I will update it with a solution for everyone that comes across:
Grafana 4.5.x supports the modification of the frontend files.

1. Create the file `/public/css/grafana.custom.css` and add your custom rules.
2. Modify `/public/views/index.html` and add `<link rel="stylesheet" type="text/css" href="[[.AppSubUrl]]/public/css/grafana.custom.css" />` into the head section. Add it _after_ the other `css` file references.
3. Restart grafana to purge the cache of the index.html page (it's only read once per startup).
4. You can customize the `grafana.custom.css` and see the results without restarting the grafana service. Just refresh the browser to see the effects.

Be aware that you have to make these changes every time you update grafana, so keep the `grafana.custom.css` in a safe place.

**Comment by amc90** [comment]
This would be great - it doesn't take many columns to make the table unusable at the moment.

**Comment by rbjorklin** [comment]
Has there been any progress on this issue? Vertical headers would be really nice to have.

**Comment by kennedyjosh** [comment]
Any update here? Vertical (or perhaps angled?) headers would still be great

**Comment by nabelekt** [comment]
Definitely agree that vertical table headers are needed. We often want to display just the last values for several different series in a table. In the current table panel, that creates just one row that has to scroll pretty far horizontally. Being able to simply transpose the table would fix this.

**Comment by nabelekt** [comment]
Any updates here? Still doesn't seem possible on v9.4.1. This is my biggest gripe with Grafana. Pretty much every dashboard I build has a table where I want to show only the latest values for often many fields. Having those in rows just doesn't work nearly as well as having them in a column would.

I think the old Table panel did this fine. Unfortunately, that is no longer selectable as the panel type. A work around is to copy an old pre-existing Table panel to use on new dashboards.

**Comment by baonq-me** [comment]
+1

**Comment by natellium** [comment]
Hey there! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

I will be closing this discussion, but if you’re still interested in this request being implemented please don’t hesitate to open a [feature request](https://github.com/grafana/grafana/issues/new?assignees=&labels=&projects=&template=1-feature_requests.md). Sorry for the inconvenience, and thanks for your contributions!

cc @grafana/grafana-bi-squad
